Better absolute path handling

JunkuiZhang opened the pull request:
Closes #19866

This PR supersedes #19228, as #19228 encountered too many merge conflicts.

After some exploration, I found that for paths with the `\\?\` prefix, we can safely remove it and consistently use the clean paths in all cases. Previously, in #19228, I thought we would still need the `\\?\` prefix for IO operations to handle long paths better. However, this turns out to be unnecessary because Rust automatically manages this for us when calling IO-related APIs. For details, refer to Rust's internal function [`get_long_path`](https://github.com/rust-lang/rust/blob/017ae1b21f7be6dcdcfc95631e54bde806653a8a/library/std/src/sys/path/windows.rs#L225-L233).

Therefore, we can always store and use paths without the `\\?\` prefix.

This PR introduces a `SanitizedPath` structure, which represents a path stripped of the `\\?\` prefix. To prevent untrimmed paths from being mistakenly passed into `Worktree`, the type of `Worktree`’s `abs_path` member variable has been changed to `SanitizedPath`. 

Additionally, this PR reverts the changes of #15856 and #18726. After testing, it appears that the issues those PRs addressed can be resolved by this PR.

### Existing Issue
To keep the scope of modifications manageable, `Worktree::abs_path` has retained its current signature as `fn abs_path(&self) -> Arc<Path>`, rather than returning a `SanitizedPath`. Updating the method to return `SanitizedPath`—which may better resolve path inconsistencies—would likely introduce extensive changes similar to those in #19228.

Currently, the limitation is as follows:

```rust
let abs_path: &Arc<Path> = snapshot.abs_path();
let some_non_trimmed_path = Path::new("\\\\?\\C:\\Users\\user\\Desktop\\project"); 
// The caller performs some actions here:
some_non_trimmed_path.strip_prefix(abs_path);  // This fails
some_non_trimmed_path.starts_with(abs_path);   // This fails too
```

The final two lines will fail because `snapshot.abs_path()` returns a clean path without the `\\?\` prefix. I have identified two relevant instances that may face this issue:
- [lsp_store.rs#L3578](https://github.com/zed-industries/zed/blob/0173479d18e2526c1f9c8b25ac94ec66b992a2b2/crates/project/src/lsp_store.rs#L3578)
- [worktree.rs#L4338](https://github.com/zed-industries/zed/blob/0173479d18e2526c1f9c8b25ac94ec66b992a2b2/crates/worktree/src/worktree.rs#L4338)

Switching `Worktree::abs_path` to return `SanitizedPath` would resolve these issues but would also lead to many code changes.

Any suggestions or feedback on this approach are very welcome.

cc @SomeoneToIgnore 

Release Notes:

- N/A

JunkuiZhang committed:
remove previouse changes

JunkuiZhang committed:
init

JunkuiZhang committed:
checkpoint

JunkuiZhang committed:
remove unused

JunkuiZhang committed:
fix macOS

JunkuiZhang committed:
new function

CharlesChen0823 commented:
IMO, one unc path will introduce, might in `fs::watch`? and another might in `gpui::windows::platform`?

JunkuiZhang commented:
> IMO, one unc path will introduce, might in `fs::watch`? and another might in `gpui::windows::platform`?

I'm not entirely sure I understand your point. Could you clarify a bit more?

CharlesChen0823 commented:
Sorry for unclarify, as you see, 
>  I found that for paths with the \\?\ prefix, we can safely remove it and consistently use the clean paths in all cases. 

so, I thought, we should find where the unc path is produced? If we find the root of problem. we can deal with the root.

One corner might be, use using `zed cli`, could cause the worktree with unc path, another might be `platform` os api, will generate unc path for worktree?

JunkuiZhang commented:
> so, I thought, we should find where the unc path is produced? If we find the root of problem. we can deal with the root.

Mostly UNC path is returned by calling `canonicalize` on a `Path`. There are a few different reasons that I didn't handle this here.

> One corner might be, use using `zed cli`, could cause the worktree with unc path, another might be `platform` os api, will generate unc path for worktree?

The purpose of this PR is to address all cases.


> after reading some docs, I think, the unc path generated from
> 
> 1. [`fs::canonicalize`](https://github.com/rust-lang/rust/blob/66701c42263042f7120385725606edeb987ad4f1/library/std/src/sys/pal/windows/fs.rs#L1303),
>    In [docs](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfinalpathnamebyhandlew#remarks),  When using `VOLUME_NAME_DOS`, the string that is returned by this function uses the `"\\?\"` syntax.
> 2. [`fs::read_link`](https://github.com/rust-lang/rust/blob/66701c42263042f7120385725606edeb987ad4f1/library/std/src/sys/pal/windows/fs.rs#L539), this call will also might return unc path from reading the source code.
> 3. lsp server might return unc path, I'm not reading some lsp code, but from some issue about windows, this should be.
> 
> and from the [doc](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file), mostly, we can trim `\\?\` safely, but when the file path length is exceed MAX_PATH(260), it will not proper.

As I mentioned earlier, when using Rust's API, we can always remove the `\\\\?\\` prefix.

JunkuiZhang commented:
This PR has been sitting here for a month, so let's do a quick recap.

It leverages Rust's type system to ensure the correctness of `PathBuf` on Windows by introducing a `SanitizedPath` type and changing the `abs_path` member of `Snapshot` from `Arc<Path>` to `SanitizedPath`.  

This PR is ready to be merged at any time. However, there are a few issues that I think need to be aligned with the Zed team before moving forward:  

- To minimize the scope of changes, the `Worktree::abs_path` function still returns an `Arc<Path>` rather than a `SanitizedPath`. This could lead to potential issues:  
  ```rust
  let abs_path: Arc<Path> = snapshot.abs_path();  // which is C:\\Users\\user\\Desktop\\project
  let some_non_trimmed_path = Path::new("\\\\?\\C:\\Users\\user\\Desktop\\project\\main.rs");  
  // The caller performs some actions here:
  some_non_trimmed_path.strip_prefix(abs_path);  // This fails
  some_non_trimmed_path.starts_with(abs_path);   // This fails too
  ```
  
- @ConradIrwin  has additional ideas for handling paths, as outlined in #20587.  

For the first issue, I would lean toward changing the return type of `Worktree::abs_path` to `SanitizedPath`. While this would involve more extensive modifications, it would help prevent potentially hard-to-trace bugs.  

As for the second issue, I suggest waiting until this PR is merged and the remote functionality on Windows (#20587) is implemented. We can then revisit the discussion based on further testing.  

cc @SomeoneToIgnore

SomeoneToIgnore reviewed (commented):
I think this is a good start, as it removes a bunch of hacky trims, the workspace path change seems like a good move too, but let's do that separately.

I am not very fond of `into()` all around the code, can we either use `T::from` or whatever else more readable to convert one path into another.

Besides a better API, one great thing this PR brings , is a fix to task reruns: if I open `zed` repo in the version built from `main`, I get various errors with UNC paths during `cargo check --all --all-targets` task reruns.
This does not seem to be the case for the PR, which means things like https://github.com/zed-industries/zed/pull/21101 are fixed (presumably), and Windows build's usability is back to "normal".

Ergo, I would want to merge this, but only after Wednesday's release, to be sure no random breakage is incurred onto non-Windows code.
For that, we need a rebase.

JunkuiZhang committed:
fix merge conflicts

JunkuiZhang committed:
remove some uses of `into()`

JunkuiZhang commented:
I made the following changes:  
```rust
// Before
let abs_path = path.into();
// After
let abs_path: SanitizedPath = path.into();
```  

I did not use `SanitizedPath::from` because it would require changing the function signature from:  
```rust
fn some_func(path: impl Into<SanitizedPath>)
```  
to:  
```rust
fn some_func(path: impl Into<SanitizedPath> + AsRef<Path>)
```  

This modification could result in additional cloning in scenarios like the following:  
```rust
fn func1(path: impl Into<SanitizedPath> + AsRef<Path>) {
    ...
}

fn func2(path: impl Into<SanitizedPath> + AsRef<Path>) {
    let abs_path = SanitizedPath::from(path);
    // func1(abs_path); This will fail
    func1(abs_path.as_path()); // This introduces an extra copy
}
```  

I’d be happy to hear any thoughts or suggestions on this.

SomeoneToIgnore reviewed (approved):
I see, no worries then, this is much better now

I've just rechecked the new, rebased, version to ensure it fixes all UNC-related issues for me, so LGTM to merge after Wednesday release.

Given the decision to deal with this incrementally, is there anything keeping this PR in a draft state?

JunkuiZhang committed:
add comments and tests

JunkuiZhang commented:
> I see, no worries the, this is much better now
> 
> I've just rechecked the new, rebased, version to ensure it fixes all UNC-related issues for me, so LGTM to merge after Wednesday release.
> 
> Given the decision to deal with this incrementally, is there anything keeping this PR in a draft state?

I added some comments and tests.
